File: Client-Python/client_service.py
import sys
sys.path.insert(1, '../Contract/target/generated-sources/protobuf/python')

import grpc
import TupleSpaces_pb2 as pb2
import TupleSpaces_pb2_grpc as pb2_grpc

class ClientService:
    CLIENT_PREFIX = "Client "
    REQUEST_PUT = " -> PUT request: Tuple = "
    REQUEST_READ = " -> READ request: Tuple = "
    REQUEST_TAKE = " -> TAKE request: Tuple = "
    REQUEST_GET_TUPLE_SPACES_STATE = " -> GET_TUPLE_SPACES_STATE request"
    RESPONSE_SERVER = "Response from Server: "
    SHUTDOWN_MESSAGE = "Channel is shutting down..."
    SUCCESS = "OK"
    OPERATION_SUCCESSFUL = "SUCCESS"
    DEBUG = "[DEBUG] "


    def __init__(self, host_port: str, client_id: int, debug_flag: bool):
        self.channel = grpc.insecure_channel(host_port)
        self.stub = pb2_grpc.TupleSpacesStub(self.channel)
        self.client_id = client_id
        self.debug_flag = debug_flag

    def close(self):
        self.debug(self.SHUTDOWN_MESSAGE)
        self.channel.close()

    # Helper method to print debug messages
    def debug(self, debugMessage):
        if (self.debug_flag) :
            print(self.DEBUG + debugMessage, file=sys.stderr)

    def display_success(self):
        print(self.SUCCESS)

    def put(self, tuple_data: str):
        self.debug(self.CLIENT_PREFIX + str(self.client_id) + self.REQUEST_PUT + tuple_data)
        try:
            response = self.stub.put(pb2.PutRequest(clientId=self.client_id, newTuple=tuple_data))
        except grpc.RpcError:
            raise
        self.debug(self.RESPONSE_SERVER + self.OPERATION_SUCCESSFUL)
        self.display_success()

    def read(self, tuple_data: str):
        self.debug(self.CLIENT_PREFIX + str(self.client_id) + self.REQUEST_READ + tuple_data)
        try:
            response = self.stub.read(pb2.ReadRequest(clientId=self.client_id, searchPattern=tuple_data))
        except grpc.RpcError:
            raise
        self.debug(self.RESPONSE_SERVER + self.OPERATION_SUCCESSFUL)
        self.display_success()
        return response.result

    def take(self, tuple_data: str):
        self.debug(self.CLIENT_PREFIX + str(self.client_id) + self.REQUEST_TAKE + tuple_data)
        try:
            response = self.stub.take(pb2.TakeRequest(clientId=self.client_id, searchPattern=tuple_data))
        except grpc.RpcError:
            raise
        self.debug(self.RESPONSE_SERVER + self.OPERATION_SUCCESSFUL)
        self.display_success()
        return response.result

    def get_tuple_spaces_state(self) -> list[str]:
        self.debug(self.CLIENT_PREFIX + str(self.client_id) + self.REQUEST_GET_TUPLE_SPACES_STATE)
        response = self.stub.getTupleSpacesState(pb2.getTupleSpacesStateRequest(clientId=self.client_id))
        self.debug(self.RESPONSE_SERVER + self.OPERATION_SUCCESSFUL)
        self.display_success()
        return response.tuple

File: Client-Java/src/main/java/pt/ulisboa/tecnico/tuplespaces/client/grpc/ClientService.java
package pt.ulisboa.tecnico.tuplespaces.client.grpc;

import java.util.List;

import io.grpc.ManagedChannel;
import io.grpc.ManagedChannelBuilder;
import io.grpc.Metadata;
import io.grpc.StatusRuntimeException;
import io.grpc.stub.MetadataUtils;
import pt.ulisboa.tecnico.tuplespaces.centralized.contract.PutRequest;
import pt.ulisboa.tecnico.tuplespaces.centralized.contract.PutResponse;
import pt.ulisboa.tecnico.tuplespaces.centralized.contract.ReadRequest;
import pt.ulisboa.tecnico.tuplespaces.centralized.contract.ReadResponse;
import pt.ulisboa.tecnico.tuplespaces.centralized.contract.TakeRequest;
import pt.ulisboa.tecnico.tuplespaces.centralized.contract.TakeResponse;
import pt.ulisboa.tecnico.tuplespaces.centralized.contract.TupleSpacesGrpc;
import pt.ulisboa.tecnico.tuplespaces.centralized.contract.getTupleSpacesStateRequest;
import pt.ulisboa.tecnico.tuplespaces.centralized.contract.getTupleSpacesStateResponse;

public class ClientService {

    private static final String CLIENT_PREFIX = "Client ";
    private static final String REQUEST_PUT = " -> PUT request: Tuple = ";
    private static final String REQUEST_READ = " -> READ request: Tuple = ";
    private static final String REQUEST_TAKE = " -> TAKE request: Tuple = ";
    private static final String REQUEST_GET_TUPLE_SPACES_STATE = " -> GET_TUPLE_SPACES_STATE request";
    private static final String REQUEST_DELAY = " -> requested the servers to apply the following delays: ";
    private static final String RESPONSE_SERVER = "Response from Server: ";
    private static final String SHUTDOWN_MESSAGE = "Channel is shutting down...";
    private static final String SUCCESS = "OK";
    private static final String OPERATION_SUCCESSFUL = "SUCCESS";
    private static final String DEBUG = "[DEBUG] ";
    private static final Metadata.Key<String> DELAY_KEY_1 = Metadata.Key.of("delay1", Metadata.ASCII_STRING_MARSHALLER);
    private static final Metadata.Key<String> DELAY_KEY_2 = Metadata.Key.of("delay2", Metadata.ASCII_STRING_MARSHALLER);
    private static final Metadata.Key<String> DELAY_KEY_3 = Metadata.Key.of("delay3", Metadata.ASCII_STRING_MARSHALLER);
    
    private final ManagedChannel channel;
    private final TupleSpacesGrpc.TupleSpacesBlockingStub blockingStub;
    int clientId;
    private boolean debugFlag;

    public ClientService(String hostPort, int clientId, boolean debugFlag) {
		this.channel = ManagedChannelBuilder.forTarget(hostPort).usePlaintext().build();
        this.blockingStub = TupleSpacesGrpc.newBlockingStub(this.channel);
        this.clientId = clientId;
        this.debugFlag = debugFlag;
    }

    // Getter to access stub
    public TupleSpacesGrpc.TupleSpacesBlockingStub getBlockingStub() {
        return this.blockingStub;
    }

    // Getter to access client ID
    public int getClientId() {
        return this.clientId;
    }

    // Shutdown method to properly close the channel
    public void shutdown() {
        debug(SHUTDOWN_MESSAGE);
        this.channel.shutdown();
    }

    // Helper method to print debug messages
    public void debug(String debugMessage) {
        if (this.debugFlag) {
            System.err.println(DEBUG + debugMessage);
        }
    }

    // Displays the operation success outcome
    private void displaySuccess() {
        System.out.println(SUCCESS);
    }

    // Creates a Metadata object containing delay values for the replicas
    private Metadata createMetadata(String delayReplica1, String delayReplica2, String delayReplica3) {
        Metadata metadata = new Metadata();
        metadata.put(DELAY_KEY_1, delayReplica1);
        metadata.put(DELAY_KEY_2, delayReplica2);
        metadata.put(DELAY_KEY_3, delayReplica3);
        return metadata;
    }

    // Determines which gRPC stub to use based on the client input
    private TupleSpacesGrpc.TupleSpacesBlockingStub determineStub(String delayReplica1, String delayReplica2, String delayReplica3) {
        
        // If no delay value was specified by the client
        if (delayReplica1.isEmpty() && delayReplica2.isEmpty() && delayReplica3.isEmpty()) {
            return this.blockingStub;     // Returns the default blocking stub
        }
        //Otherwise, returns a stub with the new metadata
        debug(CLIENT_PREFIX + this.clientId + REQUEST_DELAY + delayReplica1 + " ; " + delayReplica2 + " ; " + delayReplica3 + " s");
        Metadata metadata = createMetadata(delayReplica1, delayReplica2, delayReplica3);
        return this.blockingStub.withInterceptors(MetadataUtils.newAttachHeadersInterceptor((metadata)));
    }

    public void put(String tuple, String delayReplica1, String delayReplica2, String delayReplica3) {

        debug(CLIENT_PREFIX + this.clientId + REQUEST_PUT + tuple);
        TupleSpacesGrpc.TupleSpacesBlockingStub stub = determineStub(delayReplica1, delayReplica2, delayReplica3);
        PutRequest request = PutRequest.newBuilder().setClientId(clientId).setNewTuple(tuple).build();
        PutResponse response = null;
        try {
            response = stub.put(request);    // Uses the determined stub, either the default one or the one with metadata attached
        } catch (StatusRuntimeException e) { // Catching exceptions from the server
            throw e;
        }
        debug(RESPONSE_SERVER + OPERATION_SUCCESSFUL); // It only reaches here if no exception occurred
        displaySuccess();
    }

    public String read(String tuple, String delayReplica1, String delayReplica2, String delayReplica3) {

        debug(CLIENT_PREFIX + this.clientId + REQUEST_READ + tuple);
        TupleSpacesGrpc.TupleSpacesBlockingStub stub = determineStub(delayReplica1, delayReplica2, delayReplica3);
        ReadRequest request = ReadRequest.newBuilder().setClientId(clientId).setSearchPattern(tuple).build();
        ReadResponse response = null;
        try {
            response = stub.read(request);   // Uses the determined stub, either the default one or the one with metadata attached
        } catch (StatusRuntimeException e) { // Catching exceptions from the server
            throw e;
        }
        debug(RESPONSE_SERVER + OPERATION_SUCCESSFUL); // It only reaches here if no exception occurred
        displaySuccess();
        return response.getResult();
    }

    public String take(String tuple, String delayReplica1, String delayReplica2, String delayReplica3) {

        debug(CLIENT_PREFIX + this.clientId + REQUEST_TAKE + tuple);
        TupleSpacesGrpc.TupleSpacesBlockingStub stub = determineStub(delayReplica1, delayReplica2, delayReplica3);
        TakeRequest request = TakeRequest.newBuilder().setClientId(clientId).setSearchPattern(tuple).build();
        TakeResponse response = null;
        try {
            response = stub.take(request);   // Uses the determined stub, either the default one or the one with metadata attached
        } catch (StatusRuntimeException e) { // Catching exceptions from the server
            throw e;
        }
        debug(RESPONSE_SERVER + OPERATION_SUCCESSFUL); // It only reaches here if no exception occurred
        displaySuccess(); 
        return response.getResult();
    }

    public List<String> getTupleSpacesState() {
        debug(CLIENT_PREFIX + this.clientId + REQUEST_GET_TUPLE_SPACES_STATE);
        getTupleSpacesStateRequest request = getTupleSpacesStateRequest.newBuilder().setClientId(clientId).build();
        getTupleSpacesStateResponse response = this.blockingStub.getTupleSpacesState(request);
        debug(RESPONSE_SERVER + OPERATION_SUCCESSFUL);
        displaySuccess(); 
        return response.getTupleList();
    }
}
